src/Workspaces/Remote/ServiceHub/Services/DiagnosticAnalyzer/DiagnosticComputer.cs

```diff
@@ -127,9 +128,15 @@ internal class DiagnosticComputer
 
             if (logPerformanceInfo && _performanceTracker != null)
             {
-                // +1 to include project itself
-                var unitCount = documentAnalysisScope != null ? 1 : _project.DocumentIds.Count + 1;
-                _performanceTracker.AddSnapshot(analysisResult.AnalyzerTelemetryInfo.ToAnalyzerPerformanceInfo(_analyzerInfoCache), unitCount);
+                // Only log telemetry snapshot is we have an active telemetry session,
+                // i.e. user has not opted out of reporting telemetry.
+                var telemetryService = _project.Solution.Workspace.Services.GetRequiredService<IWorkspaceTelemetryService>();
+                if (telemetryService.HasActiveSession)
+                {
+                    // +1 to include project itself
+                    var unitCount = documentAnalysisScope != null ? 1 : _project.DocumentIds.Count + 1;
```

Comment:
nit:

```suggestion
                    var unitCount = 1;
                    if (documentAnalysisScope != null)
                        unitCont += _project.DocumentIds.Count;
```

might be eaier to understand.  up to you.